src/main/java/org/quiltmc/loader/impl/discovery/ModResolver.java

```diff
@@ -120,143 +203,208 @@ private static IVecInt toVecInt(IntStream stream) {
+
+					/*
+					 * Plugins would most likely be passed the map of roots, and the list of causes.
+					 * Alternatively, we could do some pre-processing to identify specific cases, such as
+					 * a mod or library (or api class implementation) being missing, and then send those
+					 * out separately.
+					 *
+					 * This does mean that we'd need to remove and re-add edited clauses though when
+					 * trying to fix problems. (For example downloading a mod should process all of the
+					 * dependencies, provides, etc of that mod related to all others). 
+					 */
+
+					ModResolutionException ex = describeError(roots, causes);
+					if (ex == null) {
+						ex = fallbackErrorDescription(roots, causes);
+					}
+
+					errors.add(ex);
+
+					if (causes.isEmpty()) {
+						break;
+					} else {
+
+						boolean removedAny = false;
 
-				for (String mod : modCandidateMap.keySet()) {
-					int pos = assumptions.size();
-					assumptions = assumptions.push(0);
-					Collection<ModCandidate> candidates = modCandidateMap.get(mod);
-					boolean satisfied = false;
+						// Remove dependences and conflicts first
+						for (ModLink link : causes) {
 
-					for (ModCandidate candidate : candidates) {
-						assumptions.set(pos, candidateIntMap.get(candidate));
-						if (problem.isSatisfiable(assumptions)) {
-							satisfied = true;
+							if (link instanceof ModDep) {
+								ModDep dep = (ModDep) link;
+
+								if (!dep.validOptions.isEmpty()) {
+									continue;
+								}
+							}
+
+							if (link instanceof ModDep || link instanceof ModBreakage) {
+								if (helper.quilt_removeConstraint(link)) {
+									removedAny = true;
+									break;
+								}
+							}
+						}
+
+						// If that failed... try removing anything else
+						if (!removedAny) {
+							for (ModLink link : causes) {
+								if (helper.quilt_removeConstraint(link)) {
+									removedAny = true;
+									break;
+								}
+							}
+						}
+
+						// If that failed... stop finding more errors
+						if (!removedAny) {
 							break;
 						}
 					}
+				}
 
-					if (!satisfied) {
-						if (mandatoryMods.contains(mod)) {
-							throw new ModResolutionException("Could not resolve mod collection including mandatory mod '" + mod + "'");
-						} else {
-							assumptions = assumptions.pop();
-						}
+				if (!errors.isEmpty()) {
+					if (errors.size() == 1) {
+						throw errors.get(0);
 					}
+					ModResolutionException ex = new ModResolutionException("Found " + errors.size() + " errors while resolving mods!");
+					for (ModResolutionException error : errors) {
+						ex.addSuppressed(error);
+					}
+					throw ex;
 				}
 
-				// TODO: better error message, have a way to figure out precisely which mods are causing this
-				if (!problem.isSatisfiable(assumptions)) {
-					throw new ModResolutionException("Could not resolve mod collection due to an unknown error");
+			} catch (TimeoutException e) {
```

Comment:
Hmm what causes this to be thrown? Or does the resolver logic throw it for us?

I don't see any operations specifying how long we have to finish

Reply:
I do *not* understand sat4j well enough to explain why it can throw `TimeoutException`. From some quick searching however it looks like it can be thrown if we set a timeout (*which we don't ATM*), or if we cancel it manually by calling `ISolver.expireTimeout` (*perhaps we could expose this via a gui button alongside a progress bar in the future?*), *or* if a solver / problem is undefined in some way (sorry, I don't know if this applies to us).

Either way this can be thrown by either `DependencyHelper.hasASolution` and `DependencyHelper.why`.